src/aspire/denoising/denoiser.py

```diff
@@ -1,31 +1,42 @@
 import logging
+from abc import ABC, abstractmethod
+
+from aspire.source.image import _ImageAccessor
 
 logger = logging.getLogger(__name__)
 
 
-class Denoiser:
+class Denoiser(ABC):
     """
-    Define a base class for denoising 2D images
+    Base class for 2D image denoisers.
     """
 
     def __init__(self, src):
         """
-        Initialize an object for denoising 2D images from the image source
+        Initialize an object for denoising 2D images from `src`.
 
-        :param src: The source object of 2D images with metadata
+        :param src: `ImageSource` providing noisy images.
         """
+
         self.src = src
         self.dtype = src.dtype
-        self.nimg = src.n
+        self.n = src.n
+        self._img_accessor = _ImageAccessor(self._denoise, self.n)
 
+    @property
     def denoise(self):
```

Comment:
I'm unsure about having both `Denoiser` and `DenoisedImageSource` provide the `ImageAccessor` interface. If so, what's the difference between them, that the latter also copies the metadata of the original `ImageSource`? Why not merge them in that case?

A related pairing is that of `OrientedSource` and `CLOrient3D` where the former calls `estimate_orientations` on the latter. Maybe we can do something similar here and leave the `denoise` method as is and have `DenoisedImageSource` provide the `ImageAccessor` interface.

Reply:
`Denoiser` is an ABC. Its core functions are to define a an interface for denoisers, handle (lazily) building concrete denoisers, and boiler plate to apply only `denoise`ing on chunks.  It does seem a little lonely today because ASPIRE only has the Cov2D denoiser concrete class (so we have an API for a single class atm).

As far as the ImageAccessor, `Denoiser` previously applied the denoising method to fixed contiguous batches using 16 lines (that have had _multiple_ bugs).  The changes in this PR apply denoising to _arbitrary_ chunks in 1 pythonic line using a package wide method.

<img width="1475" alt="Screenshot 2023-09-21 at 8 08 52 AM" src="https://github.com/ComputationalCryoEM/ASPIRE-Python/assets/47759732/e93789f1-b6ce-4a32-b3ee-81afdf38e140">



`DenoisedSource` provides the full ImageSource machinery, including metadata, caching and any applicable pipelining.  There is no corespondance with OrientedSource, other than its a source wrapper calling some class's abstracted method.  In a typical workflow the output of orientation estimation is a tiny stack of rotations.  The output of denoiser called similarly would be the entire stack of images.  For that reason it is imperative it (denoiser) can operate on chunks too.

It is possible to merge them, but then we would lose any API for denoisers, and end up with different Sources for each potential denoiser in the future.

Say someone clever came along and made an improved Kalman filter variant of cov2d using additional non stationary data features, or maybe a deep learning denoiser... where would they live and how would we define them consistently without `Denoiser(ABC)`?  Merging implies they each be their own `Source` unless I'm misunderstanding your suggestion.  I think it makes more sense to have one source that expects any(and only) `Denoiser` subclasses.  LMK, thanks.

Reply:
I tried to do a good deed and  make denoiser admit images, but there is a wrinkle in that we still want to know about `indices` for CTF.

`cov2d.ctf_idx[indices]` etc.

So I think just going to stick with the plan of reverting the removal of `src`.

Reply:
Looks good.